Convert validation context to data store

tarunvijwani opened the pull request:
This PR is for moving the validation errors to a `@wordpress/data` store instead of using context

Changes made in the PR:

 - Added reducer, selectors, actions types, and actions for the validation store.
 - Added test cases for the validation store.
 - Found instances of `useValidationContext` and refactored them to use the data store.

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Manual Testing

1.
- Add a Single Product Block to a page and choose a product with multiple attributes (Hoodie is a good one from the sample products set)
- Go to the page and without selecting any attributes try to add it to your cart.
- Verify you see errors on each input.
- Select an option and ensure the error message disappears as soon as you do.
- Add to cart and verify it is successful.

2.
- Create a page with the Cart Block on it, ensure the Shipping Calculator is enabled.
- Add items to your cart and visit the Cart Block.
- Use the shipping calculator, leave the free text fields blank, ensure validation errors appear.
- Try to calculate shipping with validation errors present. Ensure the form doesn't submit.
- Fix the validation errors, ensure the form submits and the address updates.

3. 
- Add a coupon to your store.
- Add an item to your cart and go to the Cart Block
- Use the coupon input and enter an invalid coupon, try to apply it.
- Ensure an error appears.
- Type into the coupon box again and ensure the error disappears.
- Type a valid coupon code and ensure it submits and the discount is applied to your cart.

4.
- Add the Checkout Block to a page, on the Shipping Address Block set `require phone number` to true.
- In an incognito window add items to your cart and go to the Checkout Block.
- Immediately press the `Place order` button.
- Ensure the invalid fields in the form are highlighted.
- Fill in your details, enter invalid data in each field (fill and then empty again in all cases) and ensure the ones that require validation show you errors.
- You should see errors like this:
- <img width="633" alt="image" src="https://user-images.githubusercontent.com/5656702/172273145-f670551b-3cbd-4edd-8904-9712a5df9eb4.png">
- While errors are present, try to proceed with checkout. Ensure you cannot proceed by pressing the place order button until the errors are fixed.

5.
- Install the Stripe extension and set it up.
- Go to the Cart Block and enter some invalid card details in the Stripe form.
- Ensure you see validation errors on each field: 
<img width="612" alt="image" src="https://user-images.githubusercontent.com/5656702/172273393-c3cc9ea5-0087-4ecf-823c-9cee1495772e.png">

opr committed:
Add validation reducers, actions, and action types

opr committed:
Add selector for getValidationErrors

opr committed:
Export store key and register store

opr committed:
Export validation store key

opr committed:
Move TextInput files to checkout package

opr committed:
Export ValidatedTextInput from blocks-checkout package

opr committed:
Update imports of ValidatedTextInput to reflect new location

opr committed:
Use the validation wp-data store for showing error messages

opr committed:
Export getValidationError in checkout package

opr committed:
Move validation store to checkout package

opr committed:
Move ValidationInputError to blocks-checkout package

opr committed:
Only export "exposedSelectors" from validation

Commit:
Convert validation context to data store

Commit:
Fixed linting error

Commit:
Fixed linting error

Commit:
Change the validation selectors to return a function

alexflorisca commented on packages/checkout/validation/reducers.ts:
```diff
@@ -20,10 +22,59 @@ const reducer: Reducer< Record< string, FieldValidationStatus > > = (
+				! isString( action.error ) ||
+				! state.hasOwnProperty( action.error )
+			) {
+				return state;
```
Shouldn't this filter based on `hidden: false` here? Otherwise we would also show the validation errors that are set to be hidden.

opr replied:
I am not sure I follow your question, this action `SHOW_VALIDATION_ERROR` sets `hidden` to be `false` on an error in the store, so the point of this is to show hidden ones, the checks above are for ensuring the error actually exists in the store.

alexflorisca commented on packages/checkout/validation/test/reducers.js:
```diff
@@ -0,0 +1,238 @@
+	} );
+	//Checks if it display a single error
+	it( 'Show single validation error', () => {
+		const state = {
```
It would be good to add 2 more errors to the state here that have `{ hidden: false }` to test wether this action will also hide any errors that are currently visible, and only show the specified one

opr replied:
That's not the goal of this action actually, it's just to change whether a specific action is hidden or not. But good idea to add a visible one and ensure it remains visible after this action is applied to a different error!

alexflorisca commented on packages/checkout/validation/test/selectors.js:
```diff
@@ -0,0 +1,51 @@
+//Test suite for validation selectors
+describe( 'Validation selectors', () => {
+	//Checks if it get existing validation error
+	it( 'Can get the validation error', () => {
```
I would include the name of the selector we are testing in the test description in these tests to make it a bit clearer

alexflorisca commented on packages/checkout/validation/test/selectors.js:
```diff
@@ -0,0 +1,51 @@
+		expect( validationErrorID ).toEqual( `validate-error-validationError` );
+	} );
+	//Checks if state has validation error
+	it( 'Can check if state has any validation errors', () => {
```
This test would pass if I changed the `hasValidationErrors()` function to always `return true`. Let's test the other option here as well and check that it returns `false` when there are no validation errors in the state

alexflorisca commented on packages/checkout/validation/test/reducers.js:
```diff
@@ -0,0 +1,238 @@
+/**
```
Let's convert this file to TypeScript as we are moving to using TS throughout the codebase

alexflorisca commented on packages/checkout/validation/test/selectors.js:
```diff
@@ -0,0 +1,51 @@
+/**
```
Let's convert this file to TS as well

alexflorisca reviewed (changes requested):
Nice work Tarun 👏 ! Laying down some good foundations. I left a few comments inline for a few potential improvements

opr committed:
Convert reducer and selectors to TS

opr committed:
Remove superfluous comments and improve test titles

opr committed:
Test to ensure visible errors remain visible

opr committed:
Make test for hasValidationErrors more robust

opr committed:
Augment the wp-data module to include our selectors and actions

opr committed:
Removed unused `exposedSelectors` variable

opr committed:
Remove TS error because of `instanceId` on props

opr committed:
Remove unnecessary as const

opr committed:
Use function returned by getValidationError

opr committed:
Use correct selector/action names now context has been decoupled

opr committed:
hide validation error when input value changes

opr committed:
Add correct aria-describedBy now we can get error id from store

opr committed:
Clear validation error from store when component unmounts

opr committed:
Clear validation error if input is valid

opr committed:
convert ValidationInputError to TS and get correct id/error from store

opr committed:
Ensure checkout block doesn't break when there are no errors

opr committed:
Get validation data from the store instead of context

opr committed:
Update country input to remove validation context

opr committed:
Move validation store out of checkout package

opr committed:
Move TextInput and ValidationInputError back out of the checkout package

opr committed:
Remove duplicate internal styles comment

opr committed:
Remove exports that no longer exist

opr committed:
Get validation store key from block-data

opr committed:
Make attribute-select-control use validation data store

opr committed:
Export FieldValidationStatus type

opr committed:
Make combobox use validation store not context

opr committed:
Make Address use validation store not context

opr committed:
Make Address use validation store not context

opr committed:
Use hasValidationErrors selector as a function in shipping calculator

opr committed:
Remove validation context from coupon story

opr committed:
Import VALIDATION_STORE_KEY from correct location

opr committed:
Stop coupon story from erroring

opr committed:
Update useStoreCartCoupons to use validation store not context

opr committed:
Make TotalsCoupon use validation store instead of context

opr committed:
Make AddToCartFormContext use validation store not context

opr committed:
Remove ValidationContext

opr committed:
Import FieldValidationStatus from correct location

opr committed:
Import ValidatedTextInput and ValidatedTextInput from correct location

opr committed:
Remove ValidationContextProvider

opr committed:
Update components to use validation store not context

opr committed:
Update useValidation to use the data store

opr commented:
@tarunvijwani - I've been working on this and I have updated the main PR description to include testing steps for what I changes! Thanks for your initial work on this PR, it was invaluable!

tarunvijwani commented:
> @tarunvijwani - I've been working on this and I have updated the main PR description to include testing steps for what I changes! Thanks for your initial work on this PR, it was invaluable!

@opr Thank you for working on this! 👍

tarhi-saad commented on assets/js/base/components/text-input/validated-text-input.tsx:
```diff
@@ -4,33 +4,21 @@
 import TextInput from './text-input';
 import './style.scss';
+import { ValidationInputError } from '../validation-input-error';
+import { VALIDATION_STORE_KEY } from '../../../data/validation';
```
Maybe using the re-mapped path is more appropriate in this case:

```js
import { VALIDATION_STORE_KEY } from '@woocommerce/block-data';
```

tarhi-saad commented on assets/js/data/validation/index.ts:
```diff
@@ -0,0 +1,36 @@
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+registerStore< State >( STORE_KEY, {
```
`registerStore` [is deprecated](https://github.com/WordPress/gutenberg/tree/trunk/packages/data#registerstore). Let's use `register` instead!

tarhi-saad commented on assets/js/base/components/validation-input-error/index.tsx:
```diff
@@ -2,22 +2,32 @@
  * External dependencies
  */
 import PropTypes from 'prop-types';
```
We converted this file to a `jsx` file. It would be great to use `TS types` instead of `prop-types`?

tarhi-saad commented on assets/js/base/context/hooks/payment-methods/use-payment-method-interface.ts:
```diff
@@ -10,11 +10,11 @@ import { getSetting } from '@woocommerce/settings';
 import deprecated from '@wordpress/deprecated';
 import LoadingMask from '@woocommerce/base-components/loading-mask';
 import type { PaymentMethodInterface } from '@woocommerce/types';
+import { ValidationInputError } from '@woocommerce/base-components/validation-input-error';
```
This file has a TS error for `ValidationInputError` (see line 139).  We'll need to update the `ValidationInputError` in `assets/js/types/type-defs/payment-method-interface.ts` at line 78.

tarhi-saad commented on assets/js/data/validation/test/reducers.ts:
```diff
@@ -0,0 +1,247 @@
+/**
+ * Internal dependencies
+ */
+import reducer, { FieldValidationStatus } from '../reducers';
```
We have an import error: `Module '"../reducers"' has no exported member 'FieldValidationStatus'`. Let's import it from the types.ts file instead.

tarhi-saad commented on assets/js/data/validation/test/selectors.ts:
```diff
@@ -0,0 +1,56 @@
+	getValidationError,
+	hasValidationErrors,
+} from '../selectors';
+import type { FieldValidationStatus } from '../reducers';
```
We have an import error: Module '"../reducers"' has no exported member 'FieldValidationStatus'. Let's import it from the types.ts file instead.

tarhi-saad reviewed (changes requested):
Thank you, @opr, for this fantastic PR! You did an excellent job! I did review & test the changes thoroughly, and everything is working as expected. Below are some minor fixes to apply before merging our PR!

tarhi-saad committed:
Merge branch 'feature/data-store-refactor' into fix/wpdata-validation

tarhi-saad committed:
Replace the validation context in checkout-events file

tarhi-saad committed:
Use the re-mapped path for the store key import

tarhi-saad committed:
Use "register" instead of the deprecated "registerStore"

tarhi-saad committed:
Fix import error of the "FieldValidationStatus" type

tarhi-saad committed:
Use TS instead of React's "PropTypes"

tarhi-saad committed:
Fix the type of "ValidationInputError" in the "payment-method-interface"

tarhi-saad committed:
Fix error not showing on the first place order click bug

We were mutating the state in the reducer, which prevented re-rendering
on state change

tarhi-saad committed:
Fix state mutation issue in the Validation reducer
